cognitedata/reveal
# feat: Add API to list available dates, change and get selected (visible) 360 image by date (no UI)

#### Type of change

![Feat](https://img.shields.io/badge/Type-Feat-green)  

#### Jira ticket :blue_book:
https://cognitedata.atlassian.net/browse/REV-840

## Description :pencil:


**Known defects / lacking capabilities:** 
The transform of each individual revision is currently the same as whatever uniqueEventDescriptor it is grouped on in mergeDescriptors. Atm I do not know how to match files and events to get the individual transform, (it looks like they do have this, but I can't see a way to link the event with the image sets other than station_id), but doing this at a later time should be trivial as it would mostly require changes in mergeDescriptors: https://cognitedata.atlassian.net/browse/REV-859

**Added to API:**
- (Image360) `list360ImageRevisions(): Image360Revision[];` 
- (Image360) `getActiveRevision(): Image360Revision;`
- (Image360) `changeRevision(revisionId: number): Promise<void>;`
- (Image360Revision) `readonly date: Date | undefined;`
- (Image360Collection) `targetRevisionDate: Date | undefined;`

**Larger changes:**
- When we get images from CDF these are now grouped in Historical360ImageSet. This holds a list of revisions (Image360EventDescriptors) and the metadata for the Event they are grouped by.

- I introduced `Image360Revision` and split out everything image related from `Image360Entity`, so that Image360Entity now represents a station, while Image360Revision represents all sets of images that belong to this station. A huge portion of this PR is replacing entities with revisions. _Edit: Due to breaking changes I can't move the visualization box to Revisions as planed. This is now still owned by Image360Entity. The gain from this is that we only have one mesh per entity, but we loose the ability to do fade/transition between revisions as this (today) require them to both have boxes. This can be solved in other ways later tho._ 

- ~~To be able to change the revision from the Image360Entity, I am currently passing a `reloadImage` function from 
Image360ApiHelper -> Image360Facade -> Image360CollectionFactory -> Image360Entity. It would be nice to avoid this, but this is as good as I could do without going though the Cognite3DViewer.~~ Added revision as optional parameter in enter360Image.

## How has this been tested? :mag:
Using the site_id: XOM-HCU1_north_v02

I added a "Current image revision" field in the Examples UI. This only has effect when changed, and is **not** read automatically when entering a new image. I have also added some logging in Examples. This can be removed once testing is done.

For how it has been tested, see below.

## Test instructions :information_source:
1. Load the XOM-HCU1_north_v02 image set.
2. Open the debug console.
3. Navigate into any image and see that the entered revision is printed in the log and available revisions are listed.
4. Navigate to an image with more than one revision and change "Current image revision" to one of the listed ids (or one that is not listed, but nothing will happen).
7. Change "Revision date" to a really large (2000000000000) and then a really small number (2). If the image has more than one revision, the image should change to which ever is the newest (for large number) or oldest (for the small number). You can ofc test with some real times. 🕐 
8. Empty  "Revision date" and the newest image should be loaded.

Finally, test 360 images like normal:
- Enter/leave images
- Change network settings, zoom, alpha, etc.

## Checklist :ballot_box_with_check:
